Readme.md
# MooCS
##A JavaScript application interface layer for the BCS series of automation controllers.

The BCS series of automation controllers are produced by [Embedded Control Concepts][ecc]. They provide a robust platform for network addressable process automation for the hobbyist, or more commonly, the home brewer. The controllers provide an HTTP GET/POST based API for 3rd party software development and extension.

This project aims to extend that API into a JavaScript framework for interfacing with the controller. Since the included software is designed specifically with the home brewer in mind, this project aims to fill a different role: *to provide an easy to use interface layer enabling rapid development and deployment of custom applications on the platform*.

## Getting Started

In order to make use of the library, you wil need:

* A PHP-enabled web server that can reach the BCS over the network. If you are on a mac, the built-in Apache server works fine.

## The Moving Parts

The interface layer is composed of three key parts:

1. A JavaScript global namespace `MooCS` providing the tools to interface to the BCS unit, using AJAX communications
2. The excellent open source [MooTools][moo] JavaScript library
3. A PHP back end script utilizing cURL to communicate with the BCS unit

The GET/POST method the BCS API provides for communication is very well suited to AJAX, and the factory supplied web interface for the unit makes use of AJAX extensively. I wanted to use the same method for this interface layer, but XMLHttpRequest objects are not allowed to communicate outside of the domain of the serving page by the browser. This necessitated the PHP-cURL translation layer to bring the whole thing together: The interface layer communicates with the PHP script, and the PHP script in turn hands off the requests to the BCS unit, and conveys the responses back to the interface layer.

## Current Features

The project is in it's infancy. Although I have only just begun, the following features are supported:

### Please note that the library is currently read-only. I have not had opportunity to get write methods working correctly.

* *Queued Requests* - Each request to read data from the BCS device passes through a queue. If a large number of simultaneous requests are made, the BCS unit will not need to deal with them all at once. Each request is dispatched as the previous one returns.
* *Cached Structures* - The BCS uses structure files which contain a significant amount of data. The *cached structures* feature and the *queued requests* feature work together to make communicating with the device as fast and light as possible, minimizing network traffic and latency when reading consecutive values.
* *Data Read* - The interface breaks up all of the structures the BCS unit provides into logical categories ready for reading. Rather than speaking in terms of data structures the way the BCS does natively, you can instead use methods like: `myDevice.read('network', 'currentIP');`
* *Multiple BCS Support* - The interface layer can support any number of BCS target devices, each having their own instance, queue, cache, etc. It's even possible to mix and match different models of BCS!

How I Got Here
--------------
As an avid home brewer, I initially purchased a BCS controller in 2009. I found the platform to be extremely stable, and it inspired me to look beyond it's intended purpose. I'm a professional javascript application developer by trade, so I got excited about the opportunity to extend the BCS API into a technology I'm more familiar with. Long term, I hope this project will enable myself and others to deploy the BCS controllers for a variety of uses outside the brewery, including home automation, solar energy system control, and so forth.

License
-------
This work is licensed under the Creative Commons Attribution-ShareAlike 3.0 Unported License. To view a copy of this license, visit [here](http://creativecommons.org/licenses/by-sa/3.0/) or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.

Contact
-------
Contact [Carson S. Christian][ccmail] with questions or comments.


[ecc]: http://www.embeddedcontrolconcepts.com
[moo]: http://www.mootools.net
[ccmail]: mailto:cc@amplego.com

History.md
### v0.3.3 - 2012-08-21

* Clean up docs slightly, add this file.

### v0.3.2 - Sometime in 2010

* Initial "release"
* Working in read-only mode

Source/Dictionary.js
/*	Dictionary.js
	2010-10-19
	Carson S. Christian
	cchristian@moocsinterface.net
	
	Provides the translation dictionary for interpereting the BCS values.
*/
/*
	This work is licensed under the Creative Commons Attribution-ShareAlike 3.0 Unported License.
	To view a copy of this license, visit http://creativecommons.org/licenses/by-sa/3.0/
	or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
*/
/*global $$, MooCS, Element, Class, Request, typeOf, JSChart */
MooCS.Dictionary = {
	cacheable: {
		'sysname.dat': 	true,
		'ipaddr.dat': 	true,
		'ultemp.dat': 	false,
		'ulsysio.dat': 	false,
		'ucsysio.dat': 	false,
		'enetlog.dat': 	false
	},
	'460': {
		ps0: {
			name: {
				target: 'sysname.dat',
				limits: [1],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 0
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 0
			},
			state0: {
				target: 'sysname.dat',
				limits: [5],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [6],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [7],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [8],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [9],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [10],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [11],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [12],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [47],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [48],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [49],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [50],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [67],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [68],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [69],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [70],
				format: 'string'
			}
		},
		ps1: {
			name: {
				target: 'sysname.dat',
				limits: [2],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 1
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 1
			},
			state0: {
				target: 'sysname.dat',
				limits: [13],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [14],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [15],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [16],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [17],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [18],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [19],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [20],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [51],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [52],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [53],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [54],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [71],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [72],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [73],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [74],
				format: 'string'
			}
		},
		ps2: {
			name: {
				target: 'sysname.dat',
				limits: [3],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 2
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 2
			},
			state0: {
				target: 'sysname.dat',
				limits: [21],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [22],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [23],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [24],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [25],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [26],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [27],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [28],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [55],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [56],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [57],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [58],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [75],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [76],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [77],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [78],
				format: 'string'
			}
		},
		ps3: {
			name: {
				target: 'sysname.dat',
				limits: [4],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 3
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 3
			},
			state0: {
				target: 'sysname.dat',
				limits: [29],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [30],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [31],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [32],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [33],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [34],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [35],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [36],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [59],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [60],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [61],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [62],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [79],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [80],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [81],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [82],
				format: 'string'
			}
		},
		ps4: {
			name: {
				target: 'sysname.dat',
				limits: [105],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 4
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 0
			},
			state0: {
				target: 'sysname.dat',
				limits: [109],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [110],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [111],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [112],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [113],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [114],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [115],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [116],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [141],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [142],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [143],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [144],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [157],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [158],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [159],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [160],
				format: 'string'
			}
		},
		ps5: {
			name: {
				target: 'sysname.dat',
				limits: [106],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 5
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 1
			},
			state0: {
				target: 'sysname.dat',
				limits: [117],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [118],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [119],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [120],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [121],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [122],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [123],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [124],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [145],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [146],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [147],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [148],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [161],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [162],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [163],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [164],
				format: 'string'
			}
		},
		ps6: {
			name: {
				target: 'sysname.dat',
				limits: [107],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 6
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 2
			},
			state0: {
				target: 'sysname.dat',
				limits: [125],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [126],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [127],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [128],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [129],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [130],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [131],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [132],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [149],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [150],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [151],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [152],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [165],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [166],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [167],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [168],
				format: 'string'
			}
		},
		ps7: {
			name: {
				target: 'sysname.dat',
				limits: [108],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 7
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 3
			},
			state0: {
				target: 'sysname.dat',
				limits: [133],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [134],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [135],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [136],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [137],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [138],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [139],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [140],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [153],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [154],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [155],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [156],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [169],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [170],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [171],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [172],
				format: 'string'
			}
		},
		temp: {
			name0: {
				target: 'sysname.dat',
				limits: [63],
				format: 'string'
			},
			name1: {
				target: 'sysname.dat',
				limits: [64],
				format: 'string'
			},
			name2: {
				target: 'sysname.dat',
				limits: [65],
				format: 'string'
			},
			name3: {
				target: 'sysname.dat',
				limits: [66],
				format: 'string'
			},
			value0: {
				target: 'ultemp.dat',
				limits: [3],
				format: 'number',
				encoding: 'temperature-10'
			},
			value1: {
				target: 'ultemp.dat',
				limits: [4],
				format: 'number',
				encoding: 'temperature-10'
			},
			value2: {
				target: 'ultemp.dat',
				limits: [5],
				format: 'number',
				encoding: 'temperature-10'
			},
			value3: {
				target: 'ultemp.dat',
				limits: [6],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint0: {
				target: 'ultemp.dat',
				limits: [11],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint1: {
				target: 'ultemp.dat',
				limits: [12],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint2: {
				target: 'ultemp.dat',
				limits: [13],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint3: {
				target: 'ultemp.dat',
				limits: [14],
				format: 'number',
				encoding: 'temperature-10'
			}
		},
		input: {
			name0: {
				target: 'sysname.dat',
				limits: [43],
				format: 'string'
			},
			name1: {
				target: 'sysname.dat',
				limits: [44],
				format: 'string'
			},
			name2: {
				target: 'sysname.dat',
				limits: [45],
				format: 'string'
			},
			name3: {
				target: 'sysname.dat',
				limits: [46],
				format: 'string'
			},
			status0: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 0
			},
			status1: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 1
			},
			status2: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 2
			},
			status3: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 3
			}
		},
		output: {
			name0: {
				target: 'sysname.dat',
				limits: [37],
				format: 'string'
			},
			name1: {
				target: 'sysname.dat',
				limits: [38],
				format: 'string'
			},
			name2: {
				target: 'sysname.dat',
				limits: [39],
				format: 'string'
			},
			name3: {
				target: 'sysname.dat',
				limits: [40],
				format: 'string'
			},
			name4: {
				target: 'sysname.dat',
				limits: [41],
				format: 'string'
			},
			name5: {
				target: 'sysname.dat',
				limits: [42],
				format: 'string'
			},
			status0: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 0
			},
			status1: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 1
			},
			status2: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 2
			},
			status3: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 3
			},
			status4: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 4
			},
			status5: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 5
			}
		},
		network: {
			staticAddress: {
				target: 'ipaddr.dat',
				limits: [0,3],
				format: 'string',
				join: '.'
			},
			subnetMask: {
				target: 'ipaddr.dat',
				limits: [4,7],
				format: 'string',
				join: '.'
			},
			gateway: {
				target: 'ipaddr.dat',
				limits: [8,11],
				format: 'string',
				join: '.'
			},
			MAC: {
				target: 'ipaddr.dat',
				limits: [12,17],
				format: 'string',
				encoding: 'hex',
				join: ':'
			},
			currentIP: {
				target: 'ipaddr.dat',
				limits: [18,21],
				format: 'string',
				join: '.'
			},
			DHCPEnabled: {
				target: 'ipaddr.dat',
				limits: [22],
				format: 'boolean'
			},
			packetsSent: {
				target: 'enetlog.dat',
				limits: [0],
				format: 'number'
			},
			packetsReceived: {
				target: 'enetlog.dat',
				limits: [2],
				format: 'number'
			}
		},
		system: {
			firmware: {
				target: 'sysname.dat',
				limits: [0],
				format: 'string'
			}
		}
	},
// **********	*********	**********	**********	BCS-462
	'462': {
		temp: {
			name4: {
				target: 'sysname.dat',
				limits: [85],
				format: 'string'
			},
			name5: {
				target: 'sysname.dat',
				limits: [86],
				format: 'string'
			},
			name6: {
				target: 'sysname.dat',
				limits: [87],
				format: 'string'
			},
			name7: {
				target: 'sysname.dat',
				limits: [88],
				format: 'string'
			},
			value4: {
				target: 'ultemp.dat',
				limits: [7],
				format: 'number',
				encoding: 'temperature-10'
			},
			value5: {
				target: 'ultemp.dat',
				limits: [8],
				format: 'number',
				encoding: 'temperature-10'
			},
			value6: {
				target: 'ultemp.dat',
				limits: [9],
				format: 'number',
				encoding: 'temperature-10'
			},
			value7: {
				target: 'ultemp.dat',
				limits: [10],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint4: {
				target: 'ultemp.dat',
				limits: [7],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint5: {
				target: 'ultemp.dat',
				limits: [8],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint6: {
				target: 'ultemp.dat',
				limits: [9],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint7: {
				target: 'ultemp.dat',
				limits: [10],
				format: 'number',
				encoding: 'temperature-10'
			}
		},
		input: {
			name4: {
				target: 'sysname.dat',
				limits: [89],
				format: 'string'
			},
			name5: {
				target: 'sysname.dat',
				limits: [90],
				format: 'string'
			},
			name6: {
				target: 'sysname.dat',
				limits: [91],
				format: 'string'
			},
			name7: {
				target: 'sysname.dat',
				limits: [92],
				format: 'string'
			},
			status4: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 4
			},
			status5: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 5
			},
			status6: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 6
			},
			status7: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 7
			}
		},
		output: {
			name6: {
				target: 'sysname.dat',
				limits: [93],
				format: 'string'
			},
			name7: {
				target: 'sysname.dat',
				limits: [94],
				format: 'string'
			},
			name8: {
				target: 'sysname.dat',
				limits: [95],
				format: 'string'
			},
			name9: {
				target: 'sysname.dat',
				limits: [96],
				format: 'string'
			},
			name10: {
				target: 'sysname.dat',
				limits: [97],
				format: 'string'
			},
			name11: {
				target: 'sysname.dat',
				limits: [98],
				format: 'string'
			},
			name12: {
				target: 'sysname.dat',
				limits: [99],
				format: 'string'
			},
			name13: {
				target: 'sysname.dat',
				limits: [100],
				format: 'string'
			},
			name14: {
				target: 'sysname.dat',
				limits: [101],
				format: 'string'
			},
			name15: {
				target: 'sysname.dat',
				limits: [102],
				format: 'string'
			},
			name16: {
				target: 'sysname.dat',
				limits: [103],
				format: 'string'
			},
			name17: {
				target: 'sysname.dat',
				limits: [104],
				format: 'string'
			},
			status6: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 6
			},
			status7: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 7
			},
			status8: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 8
			},
			status9: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 9
			},
			status10: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 10
			},
			status11: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 11
			},
			status12: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 12
			},
			status13: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 13
			},
			status14: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 14
			},
			status15: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 15
			},
			status16: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 16
			},
			status17: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 17
			}
		}
	}
};